File: src/pratice/binarysearch/두수의합3273.java
package pratice.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class 두수의합3273 {
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static int N, X;
    public static int[] A;
    public static int  cnt = 0;

    public static void main(String[] args) throws IOException {
        N = Integer.parseInt(br.readLine());
        A = new int[N];
        String[] split = br.readLine().split(" ");
        for (int i = 0; i <N ; i++) {
            A[i] = Integer.parseInt(split[i]);
        }
        X = Integer.parseInt(br.readLine());
        Arrays.sort(A);

        for (int i = 0; i < N; i++) {
            pro(X - A[i], i);
        }
        System.out.println(cnt);
    }

    private static void pro(int num, int idx) {
        int L = idx + 1;
        int R = N - 1;
        while(L <= R){
            int mid = (L + R) / 2;
            if(A[mid] == num){
                cnt++;
                return;
            }else if(A[mid] < num){
                L = mid + 1;
            }else{
                R = mid - 1;
            }
        }
    }
}

File: src/pratice/binarysearch/두수의합3273.kt
package pratice.binarysearch

fun main() {
    val N = readln().toInt()
    val arr = readln().split(" ").map { it.toInt() }.toMutableList()
    val x = readln().toInt()
    var cnt = 0
    arr.sort()
    for (i in 0 until N - 1) {
        if (binarySearch(arr, i, x)) {
            cnt++
        }
    }
    println(cnt)
}

private fun binarySearch(arr: MutableList<Int>, idx: Int, sum: Int): Boolean {
    var L = idx + 1
    var R = arr.size - 1
    while (L <= R) {
        val mid = (L + R) / 2
        if (arr[idx] + arr[mid] == sum) {
            return true
        } else if (arr[idx] + arr[mid] < sum) {
            L = mid + 1
        } else {
            R = mid - 1
        }
    }
    return false
}
